ling_flood.py

```diff
@@ -171,19 +164,39 @@ def __init__(self):
          random and have to check queen position constantly(which might be slow), probably
          will have to be changed too with the introduction of creepy spread queens"""
         queens = self.units(QUEEN)
-        hatchery = self.units(HATCHERY)
-        if hatchery.exists:
-            hatcheries_random = self.townhalls.random
-            if self.units(SPAWNINGPOOL).ready.exists:
-                if queens.amount < hatchery.amount and hatcheries_random.is_ready \
-                    and not self.already_pending(QUEEN)\
-                    and hatcheries_random.noqueue and self.supply_left > 1:
-                    if self.can_afford(QUEEN) and not queens.closer_than(8, hatcheries_random):
-                        self.actions.append(hatcheries_random.train(QUEEN))
-            for queen in queens.idle:
-                selected = self.townhalls.closest_to(queen.position)
-                if queen.energy >= 25 and not selected.has_buff(QUEENSPAWNLARVATIMER):
-                    self.actions.append(queen(EFFECT_INJECTLARVA, selected))
+        bases = self.units(HATCHERY) | self.units(LAIR) | self.units(HIVE)
+
+        # Handle injection
+        for base in bases:
+            if base.tag in self.base_queen_relation:
+                queen = self.base_queen_relation[base.tag]
+
+                if not queen: # If queen is dead
+                    del self.base_queen_relation[base.tag]
+
+                if queen.energy >= 25 and not base.has_buff(QUEENSPAWNLARVATIMER):
+                    self.actions.append(queen(EFFECT_INJECTLARVA, base))
+
+        queen_with_no_base = queens.filter(lambda queen: queen.tag not in map(lambda unit: unit.tag, self.base_queen_relation.values()))
+        base_with_no_queen = bases.filter(lambda base: base.tag not in self.base_queen_relation.keys())
+
+        # Match queens to bases
+        for queen in queen_with_no_base:
+            if base_with_no_queen.exists:
+                base = base_with_no_queen.closest_to(queen)
+                self.base_queen_relation[base.tag] = queen_with_no_base.closest_to(base)
+                return # Make sure we don't overwrite
+            else:
+                #Spread creep
+                if queen.energy >= 25:
+                    pass
+
+        # Make more queens if needed
+        if base_with_no_queen.amount > queen_with_no_base.amount + self.already_pending(QUEEN):
+            base = base_with_no_queen.noqueue
+            if base.exists and self.can_afford(QUEEN) and self.supply_left > 2:

```

Comment:
Queens only need 2 supply to be built 
so this can be replaced `if base.exists and self.can_afford(QUEEN) and self.supply_left >= 2:`